File: src/commonTest/kotlin/com/scoreo/domain/GameTypeTest.kt
package com.scoreo.domain

import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.PlayerScore
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals

class GameTypeTest {

    private fun scores(vararg pairs: Pair<String, Int>) =
        pairs.map { PlayerScore(it.first, it.second) }

    @Test
    fun `HIGHEST_SCORE returns player with max score`() {
        val gt = GameType("1", "Belote", WinCondition.HIGHEST_SCORE)
        assertEquals(listOf("bob"), gt.computeWinners(scores("alice" to 5, "bob" to 10, "charlie" to 3)))
    }

    @Test
    fun `HIGHEST_SCORE returns all tied winners`() {
        val gt = GameType("1", "Game", WinCondition.HIGHEST_SCORE)
        val winners = gt.computeWinners(scores("alice" to 10, "bob" to 10, "charlie" to 3))
        assertEquals(setOf("alice", "bob"), winners.toSet())
    }

    @Test
    fun `LOWEST_SCORE returns player with min score`() {
        val gt = GameType("1", "Golf", WinCondition.LOWEST_SCORE)
        assertEquals(listOf("alice"), gt.computeWinners(scores("alice" to 1, "bob" to 5, "charlie" to 3)))
    }

    @Test
    fun `LOWEST_SCORE returns all tied winners`() {
        val gt = GameType("1", "Golf", WinCondition.LOWEST_SCORE)
        val winners = gt.computeWinners(scores("alice" to 1, "bob" to 1, "charlie" to 3))
        assertEquals(setOf("alice", "bob"), winners.toSet())
    }

    @Test
    fun `MANUAL returns empty list from computeWinners`() {
        val gt = GameType("1", "Custom", WinCondition.MANUAL)
        assertEquals(emptyList(), gt.computeWinners(scores("alice" to 5, "bob" to 10)))
    }

    @Test
    fun `computeWinners returns empty for empty scores`() {
        val gt = GameType("1", "Game", WinCondition.HIGHEST_SCORE)
        assertEquals(emptyList(), gt.computeWinners(emptyList()))
    }

    @Test
    fun `computeWinners with custom condition uses that condition`() {
        val gt = GameType("1", "Game", WinCondition.HIGHEST_SCORE)
        val winners = gt.computeWinners(
            scores("alice" to 10, "bob" to 5, "charlie" to 3),
            WinCondition.LOWEST_SCORE,
        )
        assertEquals(listOf("charlie"), winners)
    }

    @Test
    fun `computeWinners with HIGHEST_SCORE condition returns max`() {
        val gt = GameType("1", "Belote", WinCondition.LOWEST_SCORE)
        val winners = gt.computeWinners(
            scores("alice" to 5, "bob" to 10, "charlie" to 3),
            WinCondition.HIGHEST_SCORE,
        )
        assertEquals(listOf("bob"), winners)
    }

    @Test
    fun `computeWinners with custom condition returns all tied`() {
        val gt = GameType("1", "Game", WinCondition.HIGHEST_SCORE)
        val winners = gt.computeWinners(
            scores("alice" to 1, "bob" to 1, "charlie" to 3),
            WinCondition.LOWEST_SCORE,
        )
        assertEquals(setOf("alice", "bob"), winners.toSet())
    }
}

File: apps/scoreo/src/domain/model/gameType.test.ts
import { describe, expect, it } from 'vitest'
import { computeWinners, type GameType } from './gameType'
import type { PlayerScore } from './playerScore'

function scores(...pairs: [string, number][]): PlayerScore[] {
  return pairs.map(([playerId, score]) => ({ playerId, score }))
}

function gameType(overrides: Partial<GameType> = {}): GameType {
  return {
    id: '1',
    name: 'Game',
    winCondition: 'HIGHEST_SCORE',
    tieBreakRule: 'NONE',
    tieBreakCondition: 'HIGHEST_SCORE',
    tieBreakLabel: null,
    active: true,
    ...overrides,
  }
}

describe('computeWinners', () => {
  it('HIGHEST_SCORE returns player with max score', () => {
    const gt = gameType({ name: 'Belote', winCondition: 'HIGHEST_SCORE' })
    expect(computeWinners(gt, scores(['alice', 5], ['bob', 10], ['charlie', 3]))).toEqual(['bob'])
  })

  it('HIGHEST_SCORE returns all tied winners', () => {
    const gt = gameType({ winCondition: 'HIGHEST_SCORE' })
    const winners = computeWinners(gt, scores(['alice', 10], ['bob', 10], ['charlie', 3]))
    expect(new Set(winners)).toEqual(new Set(['alice', 'bob']))
  })

  it('LOWEST_SCORE returns player with min score', () => {
    const gt = gameType({ name: 'Golf', winCondition: 'LOWEST_SCORE' })
    expect(computeWinners(gt, scores(['alice', 1], ['bob', 5], ['charlie', 3]))).toEqual(['alice'])
  })

  it('LOWEST_SCORE returns all tied winners', () => {
    const gt = gameType({ name: 'Golf', winCondition: 'LOWEST_SCORE' })
    const winners = computeWinners(gt, scores(['alice', 1], ['bob', 1], ['charlie', 3]))
    expect(new Set(winners)).toEqual(new Set(['alice', 'bob']))
  })

  it('MANUAL returns empty list', () => {
    const gt = gameType({ name: 'Custom', winCondition: 'MANUAL' })
    expect(computeWinners(gt, scores(['alice', 5], ['bob', 10]))).toEqual([])
  })

  it('returns empty for empty scores', () => {
    const gt = gameType()
    expect(computeWinners(gt, [])).toEqual([])
  })

  it('with custom condition uses that condition', () => {
    const gt = gameType({ winCondition: 'HIGHEST_SCORE' })
    const winners = computeWinners(gt, scores(['alice', 10], ['bob', 5], ['charlie', 3]), 'LOWEST_SCORE')
    expect(winners).toEqual(['charlie'])
  })

  it('with HIGHEST_SCORE condition returns max', () => {
    const gt = gameType({ name: 'Belote', winCondition: 'LOWEST_SCORE' })
    const winners = computeWinners(gt, scores(['alice', 5], ['bob', 10], ['charlie', 3]), 'HIGHEST_SCORE')
    expect(winners).toEqual(['bob'])
  })

  it('with custom condition returns all tied', () => {
    const gt = gameType({ winCondition: 'HIGHEST_SCORE' })
    const winners = computeWinners(gt, scores(['alice', 1], ['bob', 1], ['charlie', 3]), 'LOWEST_SCORE')
    expect(new Set(winners)).toEqual(new Set(['alice', 'bob']))
  })
})
